### ISSUE TITLE
DefaultConfiguration#createCopy shouldn't share mutable listener list

### ISSUE THREAD

**Issue by dansanduleac** [gradle/gradle#6985]
`DefaultConfiguration#createCopy` [replaces](https://github.com/gradle/gradle/blob/b4d8d5d170bb4ba516e88d7fe5647e2323d791dd/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java#L792) the fresh anonymous broadcaster (`dependencyResolutionListeners`) that each new configuration gets with the one from the copied configuration.

This in conjunction with the fact that after resolve, [the listener set is cleared](https://github.com/gradle/gradle/blob/b4d8d5d170bb4ba516e88d7fe5647e2323d791dd/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java#L544), means that only one of the original and copied configrations will get a non-empty listener, depending on which one is resolved first.

### Expected Behavior

Each configuration, whether copied or not, should have its own independent `dependencyResolutionListeners` that will get called with that configuration's `ResolvableDependencies` instance.

If preserving the original listeners is important, then the listener set should be shallow-copied into a new set that the copied configuration owns.

### Current Behavior

Only one of the configurations gets to trigger the listeners for both configurations.
This leads to confusion and hacks like [what nebula.dependency-recommender does](https://github.com/nebula-plugins/nebula-gradle-interop/blob/master/src/main/kotlin/com/netflix/nebula/interop/configurations.kt) (that function called [here](https://github.com/nebula-plugins/nebula-dependency-recommender-plugin/blob/a8ad219cd8cc4c323327db85902a7b6af0bfb992/src/main/groovy/netflix/nebula/dependency/recommender/DependencyRecommendationsPlugin.java#L91)).

### Context

I am trying to `copyRecursive` the `default` configuration in all subprojects, such that I can resolve them from a root project configuration and afterwards add constraints on the original subprojects' `default` configurations.
This doesn't work with `nebula.dependency-recommender` (it never gets applied to the copied configurations) because it will never have its listener called with the correct `ResolvableDependencies` for the copied configuration.

### Steps to Reproduce (for bugs)

```gradle
configurations {
  foo
}

def fooCopy = configurations.foo.copyRecursive()
configurations.add(fooCopy)

configurations.foo.incoming.beforeResolve { incoming ->
  println("beforeResolved for foo called with: " + incoming)
}

fooCopy.incoming.beforeResolve { incoming ->
  println("beforeResolved for fooCopy called with: " + incoming)
}
```

Will print:

```
> Task :dependencies

------------------------------------------------------------
Root project
------------------------------------------------------------

foo
beforeResolved for foo called with: dependencies ':foo'
beforeResolved for fooCopy called with: dependencies ':foo'
No dependencies

fooCopy
No dependencies
```

### Your Environment

 * Build scan URL: https://gradle.com/s/az7lcdu766rfk

**Comment by bigdaz** [comment]
Thanks very much for the report. 

Ideally, the `dependency-recommender` plugin would not be modifying the dependency set inside a `beforeResolve` hook. We'd like to deprecate this behaviour, in preference of the `withDependencies` hook. (However, looking at the code it appears that the `withDependencies` actions are not copied _at all_ when a configuration is copied, so this would make things worse!)

The fact that the order of resolution determines which configuration is notified is particularly fraught.
Unfortunately, this bug is unlikely to reach the top of our priority list any time soon. Are you interested in contributing a fix?

**Comment by dansanduleac** [comment]
@bigdaz Sorry I completely missed that you responded here!
I actually contributed a fix right when I created this issue as well - #6996 - but it didn't get any attention.

I agree that relying on `beforeResolve` is kinda nasty, and `withDependencies` would be way nicer, but that's the way nebula currently works.
Why don't we also copy the `withDependencies` actions? They'd operate on each configuration's own `DependencySet` so they won't conflict. I can amend my PR to do that.

I think there is a potential pitfall with doing that, which is that if users use `withDependencies` as a hook to update some state based on the "final-ish" `DependencySet`, which would then be run once more per copy, but that sounds like a major abuse of this feature, so not sure Gradle should encourage or support that use-case.
